triples-store-client/src/main/scala/io/renku/triplesstore/client/sparql/StringInterpolator.scala

```diff
@@ -0,0 +1,91 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.renku.triplesstore.client.sparql
+
+import io.renku.jsonld.{EntityId, Property}
+import io.renku.triplesstore.client.model.{Quad, Triple, TripleObject}
+import io.renku.triplesstore.client.syntax._
+
+import java.time.{Instant, LocalDate}
+
+class StringInterpolator(private val sc: StringContext) {
+
+  import StringInterpolator.ClauseDetector
+
+  def sparql(args: Any*): Fragment = fr(args: _*)
+
+  def fr(args: Any*): Fragment = {
+    val values = args.zipWithIndex.map(makeValue)
+    Fragment(sc.s(values: _*))
+  }
+
+  private lazy val makeValue: ((Any, Int)) => String = {
+    case (a: LuceneQuery, _)      => a.asSparql.sparql
+    case (a: String, _)           => a.asTripleObject.asSparql.sparql
+    case (a: Char, _)             => a.toString.asTripleObject.asSparql.sparql
+    case (a: Float, _)            => a.asTripleObject.asSparql.sparql
+    case (a: Long, _)             => a.asTripleObject.asSparql.sparql
+    case (a: Double, _)           => a.asTripleObject.asSparql.sparql
+    case (a: Boolean, _)          => a.asTripleObject.asSparql.sparql
+    case (a: Instant, _)          => a.asTripleObject.asSparql.sparql
+    case (a: LocalDate, _)        => a.asTripleObject.asSparql.sparql
+    case (a: EntityId, _)         => a.asSparql.sparql
+    case (a: Triple, _)           => a.asSparql.sparql
+    case (a: Quad, _)             => a.asSparql.sparql
+    case (a: TripleObject, _)     => a.asSparql.sparql
+    case (a: Property, _)         => a.asSparql.sparql
+    case (a: Fragment, _)         => a.sparql
+    case (a: VarName, _)          => a.name
+    case (it: Iterable[Any], idx) => resolveIterable(it, idx)
+    case (opt: Option[Any], idx)  => opt.map(makeValue(_, idx)).getOrElse("")
+    case (arg, _)                 => sys.error(s"Unsupported value type '${arg.getClass}: $arg'")
+  }
+
+  private def resolveIterable(it: Iterable[Any], idx: Int) = {
+    import ClauseDetector.ClauseType
+    ClauseDetector
+      .detectClauseContext(sc.parts(idx))
+      .map {
+        case ClauseType.VALUES => it.map(makeValue(_, idx)).map(s => s"($s)").mkString(" ")
+        case ClauseType.IN     => it.map(makeValue(_, idx)).mkString(", ")
+      }
+      .getOrElse(sys.error("Iterable cannot be resolved in this context"))
+  }
+}
+
+private object StringInterpolator {
+
+  private object ClauseDetector {
+    private val valuesClause = ".*VALUES[\\s]*\\([\\s]*\\?\\w+[\\s]*\\)[\\s]*\\{[\\s]*$".r
+    private val inClause     = ".*IN[\\s]*\\([\\s]*$".r
+
+    sealed trait ClauseType
+    object ClauseType {
+      case object VALUES extends ClauseType
+      case object IN     extends ClauseType
+    }
+
+    def detectClauseContext(snippet: String): Option[ClauseType] = {
+      val validated = snippet.filter(c => c >= ' ' && c != '|').toUpperCase
+      if (valuesClause matches validated) Some(ClauseType.VALUES)
```

Comment:
Wondering what is when there is an `IN` and a `VALUES` in the query where the `IN` comes last. Wouldn't it return in `ClauseType.VALUES`? I thought perhaps it is enough to simply look backwards (maybe via `String#lastIndexOf` and selecting the larger index).

Reply:
I think we should be fine as the regex is matching at the end of the line. I can maybe write a test verifying that.

Reply:
Ah I think I see now, it wants to match the exact phrase until the open curly. Forget my comment above.